src/widgets/activity-group/model/services/ActivityGroupsBuildManager.ts

```diff
@@ -107,54 +123,90 @@ const createActivityGroupsBuildManager = (logger: ILogger) => {
-        event,
-      }))
-      // @todo - remove after fix on BE
-      .filter(entityEvent => !!entityEvent.entity);
-
-    const calculator = EventModel.ScheduledDateCalculator;
-
-    for (const eventActivity of entityEvents) {
-      const date = calculator.calculate(eventActivity.event);
-      eventActivity.event.scheduledAt = date;
-
-      if (!date) {
-        logger.info(
-          `[ScheduledDateCalculator.calculate]: result is null, entity|event = "${eventActivity.entity.name}|${eventActivity.event.id}"`,
-        );
-      }
-    }
+    const entityEvents = events
+      .reduce((acc, event) => {
+        const entity = idToEntity[event.entityId];
+
+        if (entity) {
+          if (enableActivityAssign) {
+            let entityAssignments: Assignment[] = [];
+            if (entity.pipelineType === ActivityPipelineType.Flow) {
+              entityAssignments = assignments.filter(
+                _assignment =>
+                  _assignment.__type === 'activityFlow' &&
+                  _assignment.activityFlowId === entity.id,
+              );
+            } else {
+              entityAssignments = assignments.filter(
+                _assignment =>
+                  _assignment.__type === 'activity' &&
+                  _assignment.activityId === entity.id,
+              );
+            }
+            if (entityAssignments.length <= 0) {
+              if (entity.autoAssign) {
+                // If there is no manual assignment for this entity,
+                // and if the entity has "auto assign: ON",
+                // then include the entity in the list.
+                acc.push({ entity, event, assignment: null });
+              } else {
+                // If there is no manual assignment for this entity,
+                // and if the entity has "auto assign: OFF",
+                // then exclude the entity from the list.
+              }
+            } else {
+              for (const assignment of entityAssignments) {
+                acc.push({ entity, event, assignment });
+              }
+            }
+          } else {
+            acc.push({ entity, event, assignment: null });
+          }
+        }
 
-    entityEvents = entityEvents.filter(x => x.event.scheduledAt);
+        return acc;
+      }, [] as EventEntity[])
+      .map(entityEvent => {
+        const date = scheduledDateCalculator.calculate(entityEvent.event);
+        entityEvent.event.scheduledAt = date;
 
-    entityEvents = entityEvents.filter(x => !x.entity.isHidden);
+        if (!date) {
+          logger.info(
+            `[ScheduledDateCalculator.calculate]: result is null, entity|event = "${entityEvent.entity.name}|${entityEvent.event.id}"`,
+          );
+        }
 
-    entityEvents = sort(entityEvents);
+        return entityEvent;
+      })
+      .filter(x => x.event.scheduledAt)
+      .filter(x => !x.entity.isHidden);
+
+    const sortedEntityEvents = sort(entityEvents);
```

Comment:
It looks like we each refactored `ActivityGroupsBuildManager` slightly differently. I think most of the differences are in how we each chose to loop through the array of events, but there's also one important **functional difference** that I think needs to be fixed in the mobile app: in the web app, if an activity is both `autoAssign: true`, _and_ has manual assignments, then the web app takes both into consideration and includes a self-report for that activity, and any additional manual assignments. It appears that in the mobile app, however, if it's discovered that the activity has manual assignments, then the `autoAssign` setting is disregarded.

Though it's true that in the admin app it's hard to have an activity that's both auto-assigned and have manual assignments (because the manual assignments panel disables choosing activities set to `autoAssign: true`), it's **not impossible**: You can set an activity to `autoAssign: false`, add manual assignments, then set it to `autoAssign: true` again.

Kenroy and I discussed this a while back and agreed it was important to anticipate this possibility and support both assignment states, both because the BE allows it, and because it may indeed becomes desirable from a requirements perspective to explicitly support it after we've seen how assignments are used out in the wild. Then if we did want to better support it, then we'd only need to make changes in the admin app, not the respondent web & mobile apps as well.

Let me know if you agree with adding support for activities that are both auto-assigned, and have manual assignments – and if so, could you make the required adjustment? 🙏🏻 

This is a more minor point but still worth noting: when I approached refactoring this function, I observed how many times the array was being reiterated and reprocessed, in a way that was both less performant and harder to read, than simply iterating through it just once via a simple `for` loop. So I converted it to the latter. I think you've reduced the number of iterations, but there's still 2 – a `reduce` followed by a `map`. I still find my `for` loop easier to follow (while I have a good feel for `reduce` now, it still sometimes hurts my brain!).

Did you reference my changes to the web app's `ActivityGroupsBuildManager` when you refactored this function, or did you feel there wasn't enough in common between the two implementations for it to be worth it?

Reply:
For reference, here's the version of the [loop in the web app](https://github.com/ChildMindInstitute/mindlogger-web-refactor/blob/26d0036cd1d0bebcd92284134a136cab9862ecb3/src/widgets/ActivityGroups/model/services/ActivityGroupsBuildManager.ts#L104-L141):

```tsx
    const eventEntities: EventEntity[] = [];
    const calculator = EventModel.ScheduledDateCalculator;

    for (const event of events) {
      const entity = idToEntity[event.entityId];
      if (!entity || entity.isHidden) continue;

      event.scheduledAt = calculator.calculate(event);
      if (!event.scheduledAt) continue;

      if (params.assignments) {
        const assignments = idToAssignments[entity.id] ?? [];

        // Add auto-assignment if enabled for this activity/flow
        if (entity.autoAssign) {
          eventEntities.push({ entity, event, targetSubject: null });
        }

        // Add any additional assignments (without duplicating possible auto-assignment)
        for (const { respondentSubject, targetSubject } of assignments) {
          const isSelfAssign = respondentSubject.id === targetSubject.id;
          if (entity.autoAssign && isSelfAssign) continue;

          eventEntities.push({
            entity,
            event,
            targetSubject: isSelfAssign ? null : targetSubject,
          });
        }
      } else {
        // Assignments disabled
        eventEntities.push({
          entity,
          event,
          targetSubject: null,
        });
      }
    }
```

Reply:
Honestly I did not reference the changes in the web app for this part of the changes, mostly because the way I did all the changes were very incremental in order to not break anything. So I didn't realize there are some functional differences between the 2 versions. But I'm happy to update my code to do the same things.

With regard to the reduce followed by map, that's also just due to the incremental nature of the changes. The map was already there originally, which I didn't originally wanted to touch. But throughout the refactor process, the stuff in the reduce got added there and things just sort of settled down to the way they are now. Plus, originally I didn't have any working tests to back me up, so I was just being extra careful to not break anything that's existing. But I'm happy to combine the 2 to make them more efficient.

Reply:
Actually for the first part about functional parity, it's completely my bad for not catching it, regardless if I referenced your code or not.

The reasons I said that is because my code is basically:

```
if (entityAssignments.length <= 0) {
  if (entity.autoAssign) {
    // action block A1
  } else {
    // action block A2
  }
} else {
  // action block B
}
```

... and had I been more careful, I would have noticed the outer `else`'s action block is missing the same `autoAssign` check.

My normal approach to coding this sort of nested conditions is to just fan out all the branches, then fill in the details:

```
if (entityAssignments.length <= 0) {
  if (entity.autoAssign) {
    ...
  } else {
    ...
  }
} else {
  if (entity.autoAssign) {
    ...
  } else {
    ...
  }
}
```

This way it's easier for me to catch any "holes" in the logical branching. But I didn't do that this time, and it immediately came back the bit me in the butt.

So anyway, I'll get this fixed asap.